module/zfs/arc.c

```diff
@@ -5193,58 +5200,65 @@ arc_get_data_buf(arc_buf_hdr_t *hdr, uint64_t size, void *tag)
 void
 arc_wait_for_eviction(uint64_t amount)
 {
-	mutex_enter(&arc_evict_lock);
-	if (arc_is_overflowing()) {
-		arc_evict_needed = B_TRUE;
-		zthr_wakeup(arc_evict_zthr);
-
-		if (amount != 0) {
-			arc_evict_waiter_t aw;
-			list_link_init(&aw.aew_node);
-			cv_init(&aw.aew_cv, NULL, CV_DEFAULT, NULL);
+	switch (arc_is_overflowing()) {
+	case 0:
+		return;
+	case 1:
+		if (!arc_evict_needed) {
+			arc_evict_needed = B_TRUE;
+			zthr_wakeup(arc_evict_zthr);
```

Comment:
we don't need the arc_evict_lock to set arc_evict_needed?  could we also change it without the lock in  arc_reduce_target_size() and arc_evict_cb()?

Reply:
This is a bit racy, but since we are not in overflow, it does not matter. arc_reduce_target_size() in addition to arc_evict_needed also changes the arc_c, potentially significantly, so I'd additional fence makes sense to me. In arc_evict_cb() the locking is needed to reliably handle arc_evict_waiters list, but since it is called from only one thread, it does not matter much to performance.

Reply:
I'm not sure if this makes sense to me. Are you saying we don't need the `arc_evict_lock` here because we're not actually going to sleep which makes the comment surrounding arc_evict_needed/arc_wait_for_eviction dance meaningless in this case? Nonetheless, we need to either document this significantly or change the consumers of `arc_evict_needed` to be consistent. My fear is that another developer would see the lack of lock here and then deem this a bug or want to remove the lock in other places.

Reply:
I've added a comment here.  Comments around arc_evict_needed/arc_wait_for_eviction are about arc_evict_waiters, not touched here.